### ISSUE TITLE
Issue adding a folder containing the `#` character

### ISSUE THREAD

**Issue by jlesage** [jlesage/docker-video-duplicate-finder#4]
Adding a folder under `Search Directories` that contains the `#` character is not handled correctly:
  - Display of the folder name stops at the `#` character.
  - Scan of such folder does not work.

**Comment by jlesage** [comment]
Issue also added to the original project: https://github.com/0x90d/videoduplicatefinder/issues/398

**Comment by Maltragor** [comment]
I had tried something for fun and found out the following:

Via a detour I came across the comment: https://github.com/AvaloniaUI/Avalonia/issues/9383#issuecomment-1306973957
_(I had the problem described in the issue when testing in an older Mint version, which does not matter here)._

I then changed the one dialog to `StorageProvider.OpenFolderPickerAsync` as a test:

```C#
public ReactiveCommand<Unit, Unit> AddIncludesToListCommand => ReactiveCommand.CreateFromTask(async () => {
#if false
        /// Original
	var result = await Utils.PickerDialogUtils.OpenDialogPicker(
		new FolderPickerOpenOptions() {
			AllowMultiple = true,
			Title = "Select folder"
		});
		if (result == null || result.Count == 0) return;
		foreach (var item in result) {
			if (!SettingsFile.Instance.Includes.Contains(item))
				SettingsFile.Instance.Includes.Add(item);
		}
#else
        /// Modification
	var result = await ApplicationHelpers.MainWindow.StorageProvider.OpenFolderPickerAsync(
		new FolderPickerOpenOptions() {
                        AllowMultiple = true,
			Title = "Select folder"
                });
		if (result == null || result.Count == 0) return;
		foreach (var item in result) {
			if (item.TryGetUri(out var path) && !SettingsFile.Instance.Includes.Contains(path.LocalPath)) 
				SettingsFile.Instance.Includes.Add(path.LocalPath);
                               // This does NOT fix the bug yet, because LocalPath only returns the truncated path.
		}
#endif
});
```

The change to `StorageProvider.OpenFolderPickerAsync` also had the consequence that no strings are returned but an object which provides the paths as [Uri](https://learn.microsoft.com/en-us/dotnet/api/system.uri?view=net-7.0) instances.
And here you could see what probably also happens with `PickerDialogUtils.OpenDialogPicker`.

The complete Uri string still contains e.g. `/videos/test#dir`.
But the `#` character separates the trailing part, so it becomes the `Fragment`. Methods like `Uri.LocalPath()` therefore only return `/videos/test`.

But if you switch to `StorageProvider.OpenFolderPickerAsync`, you would at least have the possibility to get the complete path from the Uri instance. Maybe there are other ways to get the path besides `TryGetUri()`... 
But I didn't have a closer look at all this. I just wanted to know if the case is hopeless or if it is possible to get the complete path.

**Comment by jlesage** [comment]
Thanks for this, this is very useful information!

The way `Utils.PickerDialogUtils.OpenDialogPicker` is implemented is very similar to your modification and it actually does use `StorageProvider.OpenFolderPickerAsync`.  Also, you correctly found the problem: `TryGetUri()` builds an `Uri` object, but part of the file name is parsed as a `Fragment`.  Looks like this behaviour might not be the same on Windows and Linux, since the problem is not seen by the author of VDF.

So there is definitely something to do with this problem.  I tried a few things and came up with something that work... but I'm not sure if it's a solution or a hack :). I will do a PR on the project so this can be discussed with the author.

**Comment by jlesage** [comment]
PR submitted: https://github.com/0x90d/videoduplicatefinder/pull/400

**Comment by Maltragor** [comment]
Oops, now I realize that `Utils.PickerDialogUtils `is a wrapper class from VDF that uses `StorageProvider.OpenFolderPickerAsync`. I hadn't looked that far and assumed it was the older Avalonia API ;-)

It's not just the # sign, by the way. If escape sequences occur in the path, they would be converted by LocalPath(). `/test %61 %23 xyz` becomes `/test a # xyz`.

On Windows, the # character also only works because there is no fragment splitting for `new Uri(new Uri("file://"), "c:\\abc # def")` unlike `new Uri(new Uri("file://"), "/abc # def")`. 

In Avalonia 11.0.0-preview4, at least Windows and Linux: 

- `OpenFolderPickerAsync()` of the corresponding storage provider returns `files.Select(f => new BclStorageFolder(new DirectoryInfo(f))).ToArray()`. 

- `BclStorageFolder.TryGetUri()` returns `uri = Path.IsPathRooted(_directoryInfo.FullName) ? new Uri(new Uri("file://"), _directoryInfo.FullName) : new Uri(_directoryInfo.FullName, UriKind.Relative)`.

- And `new Uri(new Uri("file://"), _directoryInfo.FullName) ` causes the problems.

In 11.0.0-preview5 some changes were made to BclStorageFolder around the Uri conversion. I assume that the errors have been fixed.
Unfortunately from preview4 to preview5 some API changes have been made, so that also changes to VDF are necessary to upgrade.

**Comment by jlesage** [comment]
> It's not just the # sign, by the way. If escape sequences occur in the path, they would be converted by LocalPath().

Yes you are right...

> On Windows, the # character also only works because there is no fragment splitting for new Uri(new Uri("file://"), "c:\\abc # def") unlike new Uri(new Uri("file://"), "/abc # def").

So you mean that because the path is starting with `c:\\`, no fragment splitting is done ?

> And new Uri(new Uri("file://"), _directoryInfo.FullName)  causes the problems.

Again, because of the previous reason ?

> Unfortunately from preview4 to preview5 some API changes have been made, so that also changes to VDF are necessary to upgrade.

That might be the next step to try.  Else, if you check my PR, a workaround would be to use always `OriginalString`, even if there is no fragment.

**Comment by Maltragor** [comment]
> So you mean that because the path is starting with c:\\, no fragment splitting is done 

I had wondered why there were no problems with the # on Windows and tried around a bit.
_I thought, it could have been that the OpenFolderPickerAsync() implementation of Windows returns an escaped string. But that is not the case._

I found out that 
```
new Uri(new Uri("file://"), "/abc # def").LocalPath -> "/abc "
new Uri(new Uri("file://"), "c:\\abc # def").LocalPath -> "c:\\abc # def"
new Uri(new Uri("file://"), "\\abc # def").LocalPath -> "/abc "
new Uri(new Uri("file://"), "x:/abc # def").LocalPath -> "x:\\abc # def"
new Uri(new Uri("file://"), "xx:/abc # def").LocalPath -> "/abc "
```
Based on this, a string starting with a single `letter` followed by `:` and `/` or `\\` is obviously not split at the `#`.

> Again, because of the previous reason ?

The statement should simply be:
This place in the Avalonia code causes the error.
BclStorageFolder still gets the complete correct path under both Windows and Linux, but `new Uri(new Uri("file://"), _directoryInfo.FullName)` then does not lead to the desired result.

So these are the answers to the two questions I asked myself:
Why does it work on Windows and not on Linux. _(Whereas, as it turned out, it does only partially work under Windows  -> `%`)_. And where is the bug.

In preview5 it is changed to:
`return StorageProviderHelpers.FilePathToUri(DirectoryInfo.FullName);`
```C#
 public static Uri FilePathToUri(string path)
    {
        var uriPath = new StringBuilder(path)
            .Replace("%", $"%{(int)'%':X2}")
            .Replace("[", $"%{(int)'[':X2}")
            .Replace("]", $"%{(int)']':X2}")
            .ToString();

        return new UriBuilder("file", string.Empty) { Path = uriPath }.Uri;
    }
```

>  Else, if you check my PR, a workaround would be to use always OriginalString, even if there is no fragment.

I had seen the PR. For that reason I tried under Windows why it was not split there.

**Comment by jlesage** [comment]
This is now fixed with image version 23.02.4.
